### PR TITLE
[60] Implement page guard

### PR DESCRIPTION
**Pull request by PKhing**
## What did you do

- Implement Layout Component (with stupid header and footer)
- Implement UserContext for storing user data (with loader)
anyone who wants to use user data can call `useUser()` to get user's `firstname`, `type`, `isVerified`
- Implement page guard HOC
anyone who wants to guard their page can use `withGuard()` when exporting the component
  - the first argument is the Component you want to export
  - the second argument is array of the allowed user type 
```tsx
export default Example

// allow only Actor
export default withGuard(Example, [ UserType.ACTOR ])

// allow only Admin and Casting
export default withGuard(Example, [ UserType.ADMIN, UserType.CASTING ])
```

### CODE REVIEW

File: apps/web/src/common/context/UserContext/hooks/useUserData/index.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+const useUserData = () => {
+  const [user, setUser] = React.useState<UserData | null>(null)
+  const [isLoading, setLoading] = React.useState(true)
+
+  useEffect(() => {
+    const getUser = async () => {
+      // TODO fetch user data from API
+      // will write test for this file after API call is implemented
+      setTimeout(() => {
```
Comment (miello) [inline apps/web/src/common/context/UserContext/hooks/useUserData/index.ts:14]: (Optional)
Please consider making cleanup function for setTimeout

File: apps/web/package.json [hunk 2]
```diff
@@ -17,6 +17,7 @@
     "@emotion/server": "^11.10.0",
     "@emotion/styled": "^11.10.5",
     "@mui/material": "^5.11.4",
+    "@mui/system": "^5.11.5",
```
Comment (miello) [inline apps/web/package.json:20]: Do we still need this ?

File: apps/web/src/common/components/Layout/styled.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { styled } from '@mui/material'
+
+export const PageContainer = styled('div')({
+  display: 'flex',
+  justifyContent: 'center',
+  alignItems: 'center',
+  flexDirection: 'column',
+  height: '100vh',
```
Comment (miello) [inline apps/web/src/common/components/Layout/styled.ts:8]:
```suggestion
  minHeight: '100vh',
```

File: apps/web/src/common/hoc/withGuard/index.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    if (!user) {
+      // TODO change path
+      router.replace('/login')
+      return null
+    }
+
+    if (!user.isVerified) {
+      // TODO change path
+      router.replace('/waiting')
+    }
```
Comment (miello) [inline apps/web/src/common/hoc/withGuard/index.tsx:24]:
```suggestion
       return null
    }
```

File: pnpm-lock.yaml [hunk 5]
```diff
@@ -110,6 +110,7 @@ importers:
       '@emotion/server': ^11.10.0
       '@emotion/styled': ^11.10.5
       '@mui/material': ^5.11.4
+      '@mui/system': ^5.11.5
```
Comment (miello) [inline pnpm-lock.yaml:113]: Might not necessary because of peer-dependency

**Review by miello — CHANGES_REQUESTED** [review]

File: apps/web/src/common/context/UserContext/hooks/useUserData/index.ts (hunk 1 above)
Comment (PKhing) [reply apps/web/src/common/context/UserContext/hooks/useUserData/index.ts:14]: let's ignore it since it will be changed to API call soon

**Review by miello — APPROVED** [review]

**Pull request merged**